File: 1563-stone-game-v/solution.cpp
#include <vector>
#include <numeric>
#include <algorithm>

class Solution {
public:
    int stoneGameV(std::vector<int>& stoneValue) {
        int n = stoneValue.size();
        std::vector<int> prefixSum(n + 1, 0);
        for (int i = 0; i < n; ++i) {
            prefixSum[i + 1] = prefixSum[i] + stoneValue[i];
        }

        std::vector<std::vector<int>> memo(n, std::vector<int>(n, -1));

        auto getSum = [&](int l, int r) {
            return prefixSum[r + 1] - prefixSum[l];
        };

        auto solve = [&](auto& self, int i, int j) -> int {
            if (i == j) {
                return 0;
            }
            if (memo[i][j] != -1) {
                return memo[i][j];
            }

            int maxScore = 0;
            for (int k = i; k < j; ++k) {
                int leftSum = getSum(i, k);
                int rightSum = getSum(k + 1, j);

                if (leftSum < rightSum) {
                    maxScore = std::max(maxScore, leftSum + self(self, i, k));
                } else if (leftSum > rightSum) {
                    maxScore = std::max(maxScore, rightSum + self(self, k + 1, j));
                } else {
                    int takeLeft = leftSum + self(self, i, k);
                    int takeRight = rightSum + self(self, k + 1, j);
                    maxScore = std::max({maxScore, takeLeft, takeRight});
                }
            }

            return memo[i][j] = maxScore;
        };

        return solve(solve, 0, n - 1);
    }
};

File: 1563-stone-game-v/solution.py
from functools import lru_cache
from typing import List

class Solution:
    def stoneGameV(self, stoneValue: List[int]) -> int:
        n = len(stoneValue)
        prefix_sum = [0] * (n + 1)
        for i in range(n):
            prefix_sum[i + 1] = prefix_sum[i] + stoneValue[i]

        def get_sum(l: int, r: int) -> int:
            return prefix_sum[r + 1] - prefix_sum[l]

        @lru_cache(maxsize=None)
        def solve(i: int, j: int) -> int:
            if i == j:
                return 0

            max_score = 0
            for k in range(i, j):
                left_sum = get_sum(i, k)
                right_sum = get_sum(k + 1, j)

                if left_sum < right_sum:
                    max_score = max(max_score, left_sum + solve(i, k))
                elif left_sum > right_sum:
                    max_score = max(max_score, right_sum + solve(k + 1, j))
                else:
                    take_left = left_sum + solve(i, k)
                    take_right = right_sum + solve(k + 1, j)
                    max_score = max(max_score, take_left, take_right)

            return max_score

        return solve(0, n - 1)